cmd/flexvolume/cmd/unmount.go

```diff
@@ -0,0 +1,119 @@
+	"github.com/spf13/cobra"
+	mountutil "k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/volume/util"
+)
+
+var (
+	unmountCmd = &cobra.Command{
+		Use:   "unmount",
+		Short: "Unmounts the pod volume",
+		RunE:  unmount,
+	}
+)
+
+func init() {
+	RootCmd.AddCommand(unmountCmd)
+}
+
+func unmount(cmd *cobra.Command, args []string) error {
+
+	client, err := getRPCClient()
+	if err != nil {
+		return fmt.Errorf("Rook: Error getting RPC client: %v", err)
+	}
+
+	var opts = &flexvolume.AttachOptions{
+		MountDir: args[0],
+	}
+
+	// unmounts the volume
+	log(client, fmt.Sprintf("unmounting mount dir: %s", opts.MountDir), false)
+
+	err = client.Call("FlexvolumeController.GetAttachInfoFromMountDir", opts.MountDir, &opts)
+	if err != nil {
+		log(client, fmt.Sprintf("Unmount volume at mount dir %s failed: %v", opts.MountDir, err), true)
+		return fmt.Errorf("Unmount volume at mount dir %s failed: %v", opts.MountDir, err)
+	}
+
+	mounter := getMounter()
+
+	var globalVolumeMountPath string
+	err = client.Call("FlexvolumeController.GetGlobalMountPath", opts.VolumeName, &globalVolumeMountPath)
+	if err != nil {
+		log(client, fmt.Sprintf("Detach volume %s/%s failed. Cannot get global volume mount path: %v", opts.Pool, opts.Image, err), true)
+		return fmt.Errorf("Rook: Unmount volume failed. Cannot get global volume mount path: %v", err)
+	}
+
+	var refs []string
+	err = redirectStdout(
+		client,
+		func() error {
+			refs, err = mountutil.GetMountRefs(mounter.Interface, opts.MountDir)
+			if err != nil {
+				glog.Errorf("failed to get reference mount %s", opts.MountDir)
+				return err
+			}
+
+			// Unmount pod mount dir
+			if err := util.UnmountPath(opts.MountDir, mounter.Interface); err != nil {
+				return fmt.Errorf("failed to unmount volume at %s: %+v", opts.MountDir, err)
+			}
+
+			// Remove attachment item from the CRD
+			err = client.Call("FlexvolumeController.RemoveAttachmentObject", opts, nil)
+			if err != nil {
+				log(client, fmt.Sprintf("Unmount volume %s failed: %v", opts.MountDir, err), true)
+				// Do not return error. Try detaching first. If error happens during detach, Kubernetes will retry.
+			}
+
+			// If len(refs) is 1, then all bind mounts have been removed, and the
+			// remaining reference is the global mount. It is safe to detach. Unmount global mount dir
+			if len(refs) == 1 {
+				if err := util.UnmountPath(globalVolumeMountPath, mounter.Interface); err != nil {
+					return fmt.Errorf("failed to unmount volume at %s: %+v", opts.MountDir, err)
+				}
+			}
+
+			return nil
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	if len(refs) == 1 {
+		// call detach
+		log(client, fmt.Sprintf("calling agent to detach mountDir: %s", opts.MountDir), false)
+		err = client.Call("FlexvolumeController.Detach", opts, nil)
+		if err != nil {
+			log(client, fmt.Sprintf("Detach volume from %s failed: %v", opts.MountDir, err), true)
+			return fmt.Errorf("Rook: Unmount volume failed: %v", err)
```

Comment:
i'm trying to think through the consequences of failing the detach here.  Let's say we've successfully unmounted all consumers of the volume (and therefore we've removed all attachment objects from the CRD).  If we fail to detach here, then we are left with a CRD with zero attachment objects.   We don't have any information in the CRD that tells us what node the lingering device is still attached (mapped) to.  Is that OK?

It may very well be OK because Kubernetes will ask our flex volume to retry the detach on this same node again.  From a fencing book keeping perspective, there are 0 mounts of the device left, so that is accurate too.

But our book keeping is missing **which** node the device (e.g., `/dev/rbd0`) is still mapped on.  This feels like a problem, but I'm not sure yet.  What do you think?

Reply:
If we fail during the detach (`rbd unmap`), an error is raised. Kubernetes will try again and call `unmount(mountDir)`. The plugin will call util.UnmountPath on both the mountDir and globalDir. They will result on no-op since the devices is unmounted. And then it will retry to detach the volume again. And this hope hopefully the detach doesnt fail.

I realize this logic might seems complex but this is also what it is done on all the kubernetes volume plugin. I like your thinking though :)